### DISCUSSION TITLE
Eloquent serialization with hidden (deprecated) attributes

### DISCUSSION THREAD

**Discussion by rseon** [laravel/framework#45570] [Q&A] [answered]
Hello all,

As part of a migration strategy, I had to move an attribute (called `name`) from a table to a related pivot.
First, I want to deprecate this attribute, like this :

```php
use Illuminate\Database\Eloquent\Casts\Attribute;
use Illuminate\Database\Eloquent\Model;

class User extends Model {
    public $hidden = ['name'];
    public function name(): Attribute
    {
        throw new \Exception('Attribute name is deprecated');
    }
}
```

So when I want to get this attribute with `$user->name` the exception is correctly thrown.

But when I have to serialize it (eg using `$user->toArray()`), that attribute is still present because the exception is thrown.
If I remove my attribute casting, the attribute is correclty hidden (eg if I dump it), but the exception is not thrown anymore if I call `$user->name`.

From the exception stack trace, I think the problem is with this method : `getAttributeMarkedMutatorMethods` in the trait `Illuminate\Database\Eloquent\Concerns\HasAttributes` (read the source here : https://github.com/laravel/framework/blob/9.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L2221)

As my attribute `name` is overriden by a method, the trait try to invoke it, although the attribute is hidden, so the exception is thrown.

Do you think it could be a bug, or that I misunderstood something ?

PS : In addition, I created a Laravel example in this repository : https://github.com/rseon/laravel-deprecate-attribute

**Comment by Kristories** [comment, marked as answer]
It seems like this behavior is intended, as the `hidden` property only hides attributes from being included in the array representation of the model. The `toArray` method only checks the `hidden` property and does not account for custom attribute casting, which is why the exception is thrown in your case.

You could consider using the `$appends` property instead of custom attribute casting to hide the deprecated `name` attribute in the serialized representation of your model.
